### PR TITLE
GC improvements 6: introduce batched GC

### PR DESCRIPTION
**Pull request by teh-cmc**
Makes the GC capable of dropping entire buckets in one go when the conditions are met (and they are pretty simple to meet in the common case of in-order data).

Unfortunately, I couldn't make the batched GC match -- let alone improve -- the performance of the standard GC.
I even have a branch with a parallel batched GC, and it's still slower: the overhead of the batching datastructures just kills me everytime.
For that reason, batching is disabled by default.

I still want to commit the code so as to prevent it from rotting though, so we can come back to it at a later time.
This introduces a slight performance deterioration on the non-batched path, that's fine.

### Benchmarks

Compared to `main`:
```
group                                                                       gc_improvements_0                       gc_improvements_6
-----                                                                       -----------------                       -----------------
.../plotting_dashboard/drop_at_least=0.3/default                            7.62    652.8±4.12ms 89.8 KElem/sec     1.00     85.7±1.14ms 683.8 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=256                       5.34    465.8±2.50ms 125.8 KElem/sec    1.00     87.2±0.55ms 671.9 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=512                       7.12    655.3±2.61ms 89.4 KElem/sec     1.00     92.0±1.85ms 636.8 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=1024                      12.45  1084.0±4.47ms 54.1 KElem/sec     1.00     87.1±0.40ms 672.7 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=2048                      23.63      2.1±0.02s 27.6 KElem/sec     1.00     89.9±1.13ms 652.0 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=256/gc_batching=true                                              1.00     98.6±0.82ms 594.2 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=512/gc_batching=true                                              1.00     94.6±1.26ms 619.3 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=1024/gc_batching=true                                             1.00     93.2±1.29ms 628.9 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=2048/gc_batching=true                                             1.00     94.3±1.43ms 621.1 KElem/sec
.../timeless_logs/drop_at_least=0.3/default                                 33.30      2.4±0.03s 24.4 KElem/sec     1.00     72.2±2.46ms 811.4 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=256                            35.16      2.5±0.08s 23.5 KElem/sec     1.00     71.1±2.31ms 824.5 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=512                            35.08      2.4±0.02s 24.5 KElem/sec     1.00     68.1±1.20ms 859.9 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=1024                           36.86      2.4±0.05s 24.2 KElem/sec     1.00     65.7±0.87ms 891.4 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=2048                           35.99      2.4±0.03s 24.1 KElem/sec     1.00     67.7±1.33ms 865.9 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=256/gc_batching=true                                                   1.00     68.7±1.40ms 853.1 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=512/gc_batching=true                                                   1.00     67.3±0.32ms 870.8 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=1024/gc_batching=true                                                  1.00     67.7±1.21ms 865.2 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=2048/gc_batching=true                                                  1.00     67.6±1.31ms 866.6 KElem/sec
```

Compared to previous PR:
```
group                                                                       gc_improvements_5                       gc_improvements_6
-----                                                                       -----------------                       -----------------
.../plotting_dashboard/drop_at_least=0.3/default                            1.00     81.4±0.94ms 720.0 KElem/sec    1.05     85.7±1.14ms 683.8 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=256                       1.00     84.0±0.50ms 697.8 KElem/sec    1.04     87.2±0.55ms 671.9 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=512                       1.00     82.5±1.33ms 710.0 KElem/sec    1.11     92.0±1.85ms 636.8 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=1024                      1.00     83.4±1.16ms 702.9 KElem/sec    1.04     87.1±0.40ms 672.7 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=2048                      1.00     83.7±0.61ms 700.0 KElem/sec    1.07     89.9±1.13ms 652.0 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=256/gc_batching=true                                              1.00     98.6±0.82ms 594.2 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=512/gc_batching=true                                              1.00     94.6±1.26ms 619.3 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=1024/gc_batching=true                                             1.00     93.2±1.29ms 628.9 KElem/sec
.../plotting_dashboard/drop_at_least=0.3/bucketsz=2048/gc_batching=true                                             1.00     94.3±1.43ms 621.1 KElem/sec
.../timeless_logs/drop_at_least=0.3/default                                 1.00     66.8±0.85ms 877.3 KElem/sec    1.08     72.2±2.46ms 811.4 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=256                            1.00     67.5±1.43ms 868.2 KElem/sec    1.05     71.1±2.31ms 824.5 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=512                            1.00     67.4±1.40ms 869.4 KElem/sec    1.01     68.1±1.20ms 859.9 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=1024                           1.03     67.5±2.21ms 867.5 KElem/sec    1.00     65.7±0.87ms 891.4 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=2048                           1.00     67.8±1.86ms 863.9 KElem/sec    1.00     67.7±1.33ms 865.9 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=1024/gc_batching=true                                                  1.00     67.7±1.21ms 865.2 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=2048/gc_batching=true                                                  1.00     67.6±1.31ms 866.6 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=256/gc_batching=true                                                   1.00     68.7±1.40ms 853.1 KElem/sec
.../timeless_logs/drop_at_least=0.3/bucketsz=512/gc_batching=true                                                   1.00     67.3±0.32ms 870.8 KElem/sec
```

---

Part of the GC improvements series:
- #4394
- #4395
- #4396
- #4397
- #4398
- #4399
- #4400
- #4401

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [app.rerun.io](https://app.rerun.io/pr/4394) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4394)
- [Docs preview](https://rerun.io/preview/4ce377663695c3436c1bc201f75f846a3ee9d473/docs) 
- [Examples preview](https://rerun.io/preview/4ce377663695c3436c1bc201f75f846a3ee9d473/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_arrow_store/src/store_gc.rs [hunk 1]
```diff
@@ -222,77 +233,213 @@ impl DataStore {
         // 2. Find all tables that potentially hold data associated with that `RowId`
         // 3. Drop the associated row and account for the space we got back
 
-        for (&row_id, (timepoint, entity_path_hash)) in &self.metadata_registry.registry {
+        let batch_size = (self.config.indexed_bucket_num_rows as usize).saturating_mul(2);
```
Comment (Wumpf) [inline crates/re_arrow_store/src/store_gc.rs:236]: I think I puzzled myself together by now how we accumulate these batches and then drop them in bulk, but a few guiding comments on the steps involved would have been useful :)

**Review by Wumpf — APPROVED** [review]
lgtm, but you didn't lie, this one was a bit tougher!

File: crates/re_arrow_store/src/store.rs [hunk 2]
```diff
@@ -446,6 +445,38 @@ impl IndexedTable {
...
+                buckets,
+                all_components: _, // keep the history on purpose
+                buckets_num_rows,
+                buckets_size_bytes,
+            } = self;
+
+            let bucket = IndexedBucket::new(*cluster_key, *timeline);
+            let size_bytes = bucket.total_size_bytes();
+
+            *buckets = [(i64::MIN.into(), bucket)].into();
```
[the remark this answers is not in the document]
Comment (teh-cmc) [reply crates/re_arrow_store/src/store.rs:469]: Ha yes that's a thing nowadays

File: crates/re_arrow_store/src/store.rs [hunk 3]
```diff
@@ -446,6 +445,38 @@ impl IndexedTable {
...
+            *buckets_num_rows = 0;
+            *buckets_size_bytes = size_bytes;
+        }
+
+        // NOTE: Make sure the first bucket is responsible for `-∞`, which might or might not be
+        // the case now if we've been moving buckets around.
+        if let Some((_, bucket)) = self.buckets.pop_first() {
```
[the remark this answers is not in the document]
Comment (teh-cmc) [reply crates/re_arrow_store/src/store.rs:476]: Why not, I was just being overly cautious

File: crates/re_arrow_store/src/store_gc.rs (hunk 1 above)
Comment (teh-cmc) [reply crates/re_arrow_store/src/store_gc.rs:236]: Oh yes, as I ended up disabling it by default I completely forgot to explain what's going on :no_mouth:

**Pull request merged**